homeassistant/components/teslemetry/sensor.py

```diff
@@ -0,0 +1,446 @@
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:transmission-tower",
+    ),
+    SensorEntityDescription(
+        key="grid_services_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:transmission-tower",
+    ),
+    SensorEntityDescription(
+        key="generator_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:generator-stationary",
+    ),
+)
+
+WALL_CONNECTOR_DESCRIPTIONS: tuple[SensorEntityDescription, ...] = (
+    SensorEntityDescription(
+        key="wall_connector_state",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        icon="mdi:ev-station",
+    ),
+    SensorEntityDescription(
+        key="wall_connector_fault_state",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        icon="mdi:ev-station",
+    ),
+    SensorEntityDescription(
+        key="wall_connector_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:ev-station",
+    ),
+    SensorEntityDescription(
+        key="vin",
+        icon="mdi:car-electric",
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Teslemetry sensor platform from a config entry."""
+    data = hass.data[DOMAIN][entry.entry_id]
+
+    # Add vehicles
+    async_add_entities(
+        TeslemetryVehicleSensorEntity(vehicle, description)
+        for vehicle in data.vehicles
+        for description in VEHICLE_DESCRIPTIONS
+    )
+
+    # Add energy sites
+    async_add_entities(
+        TeslemetryEnergySensorEntity(energysite, description)
+        for energysite in data.energysites
+        for description in ENERGY_DESCRIPTIONS
+        if description.key in energysite.coordinator.data
+    )
+
+    # Add wall connectors
+    async_add_entities(
+        TeslemetryWallConnectorSensorEntity(energysite, din, description)
+        for energysite in data.energysites
+        for din in energysite.coordinator.data.get("wall_connectors", {})
+        for description in WALL_CONNECTOR_DESCRIPTIONS
+    )
```

Comment:
Only call `add_entities` once

Reply:
Done, using itertools.chain to join them up.